File: 44_wildcardMaching/wildcard_matching.py
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s:str
        :type p:str
        :rtype :bool
        """
        idx_s, idx_p = 0, 0
        for char_s in s:
            if idx_p < len(p):
                char_p = p[idx_p]
                if char_p == '?':
                    pass
                elif char_p == '*':
                    for t_p in p[idx_p:]:
                        if t_p != '*':
                            break
                        idx_p += 1
                    for i in xrange(0, len(s) - idx_s + 1):
                        if self.isMatch(s[idx_s+i:], p[idx_p:]):
                            return True
                    return False
                else:
                    if char_p != char_s:
                        return False
            else:
                return False

            idx_s += 1
            idx_p += 1

        if p[idx_p:] != '':
            for char in p[idx_p:]:
                if char != '*':
                    return False
            return True
        return True

File: 44_wildcardMaching/wildcard_matching.c
void cmp_until_diff(char **s, char **p) {
    while('*' != **p && **s && **p && (**s == **p || '?' == **p)) {
        (*s) += 1;
        (*p) += 1;
    }
}

int isMatch(char* s, char* p) {

    if('\0' == *s) {
        while(*p) {
            if('*' != *p)
                return 0;
            ++p;
        }
        return 1;
    }

    if('?' == *p) {
        return isMatch(++s, ++p);
    } else if('*' == *p) {
        while(*p) {
            if('*' == *p)
                ++p;
            else
                break;
        }
        while(*s) {
            char *p_s=s, *p_p=p;
            cmp_until_diff(&p_s, &p_p);

            if('\0' == *p_p && '\0' == *p_s) {
                return 1;
            }

            if ('*' == *p_p && isMatch(p_s, p_p)) {
                return 1;
            } else {
                ++s;
            }
        }
        return isMatch(s, p);

    } else if(*p == *s) {
        cmp_until_diff(&s, &p);
        return isMatch(s, p);
    }
    return 0;
}
